### PR TITLE
Template for Interactables and Interactors

### PR DESCRIPTION
**Pull request by seanlowjk**
# Summary
As discussed earlier this evening, we have decided to make `IInteractable` and `IInteractor` interfaces to help out with our main gameplay logic. 

## What has been done by us
Edits have been made such that we have an **association class** called `PlayerHandler` which handles the logic from the `IInteractable` objects. However, I am unable to make `PlayerMovement`, `PlayerAnimation` and `PowerUpActor` scripts to be a component or child under `PlayerHandler` as there are dependencies needed by the controller. 

## Other things to note: 
We have edited the logic such that each power up **should have their own** buff duration, and this will be passed down eventually to the `PowerUpActor`. 

Do leave comments on how we can improve this better, as we would be using this logic mainly for most of our gameplay elements.

### CODE REVIEW

File: Assets/Scripts/Interactions/IInteractable.cs [hunk 1]
```diff
@@ -0,0 +1,4 @@
+﻿public interface IInteractable
+{
+    void Interact(PlayerHandler interactor);
```
Comment (tcgj) [inline Assets/Scripts/Interactions/IInteractable.cs]: This should actually take in the ICanInteract interface

File: Assets/Scripts/Interactions/IInteractor.cs [hunk 2]
```diff
@@ -0,0 +1,4 @@
+﻿public interface IInteractor
+{
+    void HandleInteraction(IInteractable interactable);
+}
```
Comment (tcgj) [inline Assets/Scripts/Interactions/IInteractor.cs]: if I'm not wrong we wanted ICanInteract LOL. And also, the interface should probably be empty

Comment (tcgj) [reply Assets/Scripts/Interactions/IInteractor.cs]: because it seems as though this is a direct circular dependency

File: Assets/Scripts/PlayerController/PlayerHandler.cs [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    private PowerupActor m_PowerupActor;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        m_PlayerAnimation = GetComponent<PlayerAnimation>();
+        m_PlayerMovement = GetComponent<PlayerMovement>();
+        m_PowerupActor = GetComponent<PowerupActor>();
+    }
+
```
Comment (tcgj) [inline Assets/Scripts/PlayerController/PlayerHandler.cs]: This class should actually just be the PlayerNetworkHandler renamed. I can rename it to PlayerActor instead

File: Assets/Scripts/PlayerController/PlayerHandler.cs [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    public void HandleInteraction(IInteractable interactable) {
+        if (interactable is JumpPowerUp) {
+            m_PowerupActor.JumpHigher(((JumpPowerUp) interactable).GetBuffDuration());
+        } 
+
+        if (interactable is SpeedPowerUp) {
+            m_PowerupActor.GoFaster(((SpeedPowerUp) interactable).GetBuffDuration());
+        }
+    }
```
Comment (tcgj) [inline Assets/Scripts/PlayerController/PlayerHandler.cs]: Like mentioned before, this class should handle the interactions themselves, so they should actually just check for trigger/distance etc and then call the interact method of the interactable associated to it.

File: Assets/Scripts/Interactions/IInteractor.cs (hunk 2 above)
Comment (seanlowjk) [reply Assets/Scripts/Interactions/IInteractor.cs]: Ok sures will fix it!

File: Assets/Scripts/Interactions/IInteractable.cs (hunk 1 above)
Comment (seanlowjk) [reply Assets/Scripts/Interactions/IInteractable.cs]: Yea Put in wrong type for parameters YIKES

File: Assets/Scripts/PlayerController/PlayerHandler.cs (hunk 3 above)
Comment (seanlowjk) [reply Assets/Scripts/PlayerController/PlayerHandler.cs]: Ok!

File: Assets/Scripts/PlayerController/PlayerHandler.cs (hunk 4 above)
Comment (seanlowjk) [reply Assets/Scripts/PlayerController/PlayerHandler.cs]: Ok swee understooded

File: Assets/Scripts/PlayerController/PlayerHandler.cs [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    void OnTriggerEnter(Collider c)
+    {
+        IInteractable interactingObject = c.GetComponent<IInteractable>();
+        if (interactingObject != null)
+        {
+            interactingObject.Interact(this);
+        }
+    }
+
+    public void BoostJump()
```
Comment (heze8) [inline Assets/Scripts/PlayerController/PlayerHandler.cs]: Yeah, like terrence said, we don't want this code in the player handler class. It shouldn't know what the powerups do. The powerups should be the one applying the logic, and the playerhandler will only know that its something to interact with.

File: Assets/Scripts/PowerUps/PowerUp.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+using UnityEngine;
+
+public abstract class PowerUp : MonoBehaviour, IInteractable
+{
+    protected void PlayPickUpSound() 
+    {
+        AudioManager.m_Instance.PlaySound("MAGIC_Powerup", 1.0f, 1.2f);
+    }
+
+    public void Interact(ICanInteract interactor)
```
Comment (heze8) [inline Assets/Scripts/PowerUps/PowerUp.cs]: I think the interact method should contain all the common behaviour called by all powerup classes, but to implement specific behaviour for each powerup it can call a virtual method that has not been implemented.

File: Assets/Scripts/PlayerController/PlayerHandler.cs (hunk 5 above)
Comment (seanlowjk) [reply Assets/Scripts/PlayerController/PlayerHandler.cs]: Ok Sures will fix that in the next commit

File: Assets/Scripts/PowerUps/PowerUp.cs (hunk 6 above)
Comment (seanlowjk) [reply Assets/Scripts/PowerUps/PowerUp.cs]: Hmmm... in that case I'll have another abstract method to handle specific behavior

File: Assets/Scripts/PlayerController/PlayerHandler.cs [hunk 7]
```diff
@@ -0,0 +1,85 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class PlayerHandler : MonoBehaviour, ICanInteract
```
[the remark this answers is not in the document]
Comment (tcgj) [reply Assets/Scripts/PlayerController/PlayerHandler.cs]: Yep. I told Sean that I would do it instead

File: Assets/Scripts/PlayerController/PlayerHandler.cs [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+
+    private float ComputeJumpHeightModifier()
+    {
+        return m_PowerupActor.IsDoubleJump() ? 1.5f : 1.0f;
+    }
+
+    private float ComputeGravityModifier()
+    {
+        // Empirical values here
+        return m_PowerupActor.IsDoubleJump() ? 0.85f : 1.0f;
```
[the remark this answers is not in the document]
Comment (seanlowjk) [reply Assets/Scripts/PlayerController/PlayerHandler.cs]: Ok Sures, will fix them at home!

File: Assets/Scripts/PlayerController/PlayerHandler.cs [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+public class PlayerHandler : MonoBehaviour, ICanInteract
+{
+    [SerializeField]
+    private PlayerMovement m_PlayerMovement;
+
+    [SerializeField]
+    private PlayerAnimation m_PlayerAnimation;
+
+    [SerializeField]
+    private PowerupActor m_PowerupActor;
```
[the remark this answers is not in the document]
Comment (seanlowjk) [reply Assets/Scripts/PlayerController/PlayerHandler.cs]: Hmmm... in that case we would rather have all these become required components

File: Assets/Scripts/PlayerController/PlayerHandler.cs [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+        UpdateVelocity();
+    }
+
+    void OnTriggerEnter(Collider c)
+    {
+        IInteractable interactingObject = c.GetComponent<IInteractable>();
+        if (interactingObject != null)
+        {
+            interactingObject.Interact(this);
+        }
```
[the remark this answers is not in the document]
Comment (seanlowjk) [reply Assets/Scripts/PlayerController/PlayerHandler.cs]: Hmmm so what you are suggesting that we pull out his part of the code, 

and put it in a separate function, like interactWith(Collider c) ? Not quite sure what you mean by this @Eclmist

[the remark this answers is not in the document]
Comment (seanlowjk) [reply Assets/Scripts/PlayerController/PlayerHandler.cs]: Oh wait I think I got it, I'll try something out and you can advise

**Comment by tcgj** [comment]
- Merge PlayerHandler and PlayerNetworkHandler into a single class, Player.
- Remove original Forge key bindings in multiplayer menu. fixes #28 
- Add a debug flag to allow for bypassing 4-man balanced team
- Clean up powerups to only trigger interaction through player

File: Assets/Scripts/PowerUps/PowerUpBase.cs [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+            CopyPowerUpToPlayer(this, powerUp); // Use reflection to copy script onto player
+
+            powerUp.m_TimeOfActivation = Time.time;
+            powerUp.OnPowerUpActivated();
+        }
+    }
+
+    protected static void CopyPowerUpToPlayer(PowerUpBase source, PowerUpBase target)
+    {
+        FieldInfo[] sourceFields = source.GetType().GetFields(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance);
```
Comment (heze8) [inline Assets/Scripts/PowerUps/PowerUpBase.cs]: Is this related to network?

File: Assets/Scripts/PowerUps/SpeedPowerUp.cs [hunk 12]
```diff
@@ -1,21 +1,33 @@
...
-                manager.GoFaster();
-                AudioManager.m_Instance.PlaySound("MAGIC_Powerup", 1.0f, 1.0f);
-                Destroy(gameObject);
-            }
+    public override void OnPowerUpActivated()
+    {
+        PlayerMovement playerMovement = GetComponent<PlayerMovement>();
+        if (playerMovement != null)
+        {
+            playerMovement.SetExternalVelocityModifier(ComputeVelocityModifier());
```
Comment (heze8) [inline Assets/Scripts/PowerUps/SpeedPowerUp.cs]: mmm I still think that the behaviour or method should be in the power up itself, so that when we create a new powerup we don't need to create a new method in player movement?

Comment (tcgj) [reply Assets/Scripts/PowerUps/SpeedPowerUp.cs]: This is the powerup itself isn't it? I'm just modifying the player variables here

File: Assets/Scripts/PowerUps/PowerUpBase.cs (hunk 11 above)
Comment (tcgj) [reply Assets/Scripts/PowerUps/PowerUpBase.cs]: Nah. This is copying the powerup script onto the player. Allows us to destroy the powerup gameobject while keeping the behaviour in the powerup
